go/adbc/driver/bigquery/connection.go

```diff
@@ -0,0 +1,868 @@
+	if catalog == nil {
+		catalog = &c.catalog
+	}
+	sanitizedCatalog, err := sanitize(*catalog)
+	if err != nil {
+		return nil, err
+	}
+
+	if dbSchema == nil {
+		dbSchema = &c.dbSchema
+	}
+	sanitizedDbSchema, err := sanitize(*dbSchema)
+	if err != nil {
+		return nil, err
+	}
+
+	pattern, err := patternToRegexp(columnName)
+	if err != nil {
+		return nil, err
+	}
+
+	// sadly that query parameters cannot be used in place of table names
+	queryString := fmt.Sprintf("SELECT * FROM `%s`.`%s`.INFORMATION_SCHEMA.COLUMNS WHERE table_name = @tableName", sanitizedCatalog, sanitizedDbSchema)
+	query := c.client.Query(queryString)
+	query.Parameters = []bigquery.QueryParameter{
+		{
+			Name:  "tableName",
+			Value: tableName,
+		},
+	}
+	reader, _, err := newRecordReader(ctx, query, c.Alloc)
+	if err != nil {
+		return nil, err
+	}
+
+	fields := make([]arrow.Field, 0)
+	var columns map[string]int = nil
+
+	for reader.Next() && ctx.Err() == nil {
+		rec := reader.Record()
+		if columns == nil {
+			columns = make(map[string]int)
+			for i, f := range reader.Schema().Fields() {
+				columns[strings.ToUpper(f.Name)] = i
+			}
+		}
+
+		numRows := int(rec.NumRows())
+		val, ok := columns["COLUMN_NAME"]
+		if !ok {
+			return nil, adbc.Error{
+				Code: adbc.StatusInternal,
+				Msg:  "Column `COLUMN_NAME` does not exist in response",
+			}
+		}
+		fieldName := rec.Column(val)
+
+		val, ok = columns["IS_NULLABLE"]
+		if !ok {
+			return nil, adbc.Error{
+				Code: adbc.StatusInternal,
+				Msg:  "Column `IS_NULLABLE` does not exist in response",
+			}
+		}
+		fieldNullable := rec.Column(val)
+
+		val, ok = columns["DATA_TYPE"]
+		if !ok {
+			return nil, adbc.Error{
+				Code: adbc.StatusInternal,
+				Msg:  "Column `DATA_TYPE` does not exist in response",
+			}
+		}
+		fieldType := rec.Column(val)
+
+		for i := 0; i < numRows; i++ {
+			metadata := make(map[string]string)
+			metadata["PRIMARY_KEY"] = ""
+			for j := 0; j < int(rec.NumCols()); j++ {
+				col := rec.Column(j)
+				columnName := strings.ToUpper(rec.ColumnName(j))
+				switch columnName {
+				case "COLUMN_NAME", "IS_NULLABLE", "DATA_TYPE":
+					continue
+				default:
+					metadata[columnName] = strings.Clone(col.ValueStr(i))
+				}
+			}
+
+			name := fieldName.ValueStr(i)
+			if !pattern.MatchString(name) {
+				continue
+			}
+
+			nullable := strings.ToUpper(fieldNullable.ValueStr(i))
+			dataType := fieldType.ValueStr(i)
+
+			field, err := buildSchemaField(name, dataType)
+			if err != nil {
+				return nil, err
+			}
+			field.Nullable = nullable == "YES"
+			field.Metadata = arrow.MetadataFrom(metadata)
+			fields = append(fields, field)
+		}
+		fieldName.Release()
+		fieldNullable.Release()
+		fieldType.Release()
+		rec.Release()
+	}
+
+	schema := arrow.NewSchema(fields, nil)
+	return schema, nil
+}
+
+func (c *connectionImpl) Token() (*oauth2.Token, error) {
```

Comment:
does this need to be exported? Can we keep this internal?

Reply:
Sorry I might be wrong, but I was thinking that this is the required interface function for `oauth2.TokenSource`...?

```go
// A TokenSource is anything that can return a token.
type TokenSource interface {
	// Token returns a token or an error.
	// Token must be safe for concurrent use by multiple goroutines.
	// The returned Token must not be modified.
	Token() (*Token, error)
}
```